### ISSUE TITLE
DDC feature proposal: Offer a soundness checking mode.

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#30026]
It would be very useful if DDC could offer a mode where all run-time checks which are considered provably unnecessary are generated anyway, in a form where a failed test would throw a `SoundnessViolationError`.

The point is that we would know more precisely that all the run-time checks which are skipped because "they are guaranteed to succeed" actually do enjoy such a guarantee. Whenever there is a test which causes a `SoundnessViolationError` we know that we need to re-check those statically known guarantees, and this will allow us to remain very confident when we claim properties such as heap soundness.

**Comment by vsmenon** [comment]
The analyzer tells DDC where casts are needed.  Could experiment with forcing this to true:

https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/task/strong/checker.dart#L766

**Comment by eernstg** [comment]
Right, that might do it: With a deterministic semantics and tests we'd just need to check whether the "paranoid" execution has additional failures compared to the normal run (where run-time checks which are considered guaranteed-to-succeed would be omitted). In a typical situation where this helpful, the normal run would be clean and the paranoid run fails, and then we know where to look.

**Comment by jmesserly** [comment]
(moving to analyzer-strong-mode, as runtime checks are placed by the front end, which is currently Analyzer. a "paranoid" strong mode flag would work with any back end.)

**Comment by bwilkerson** [comment]
Given that analyzer won't be the front end for much longer, I've moved this to kernel.
